### PR TITLE
Add InitFinalize base class of MPI_InitFinalize 

### PR DESCRIPTION
**Pull request by Rombur**
Alternative to https://github.com/dealii/dealii/pull/14884

It creates a new base class `InitFinalize` that allows a more fine grain control over which external libraries we initialize/finalize. I have added a new function `finalize` to allow user to finalize the libraries without having the object going out-of-scope.

### CODE REVIEW

**Comment by bangerth** [comment]
```
In file included from /Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/mpi.h:22:
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:169:20: error: unknown type name 'MPI_Request'
  register_request(MPI_Request &request);
                   ^
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:175:22: error: unknown type name 'MPI_Request'
  unregister_request(MPI_Request &request);
                     ^
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:209:19: error: use of undeclared identifier 'MPI_Request'
  static std::set<MPI_Request *> requests;
                  ^
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:209:32: error: expected expression
  static std::set<MPI_Request *> requests;
                               ^
In file included from /Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/source/dofs/dof_handler.cc:20:
In file included from /Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/mpi.templates.h:22:
In file included from /Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/mpi.h:22:
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:169:20: error: unknown type name 'MPI_Request'
  register_request(MPI_Request &request);
                   ^
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:175:22: error: unknown type name 'MPI_Request'
  unregister_request(MPI_Request &request);
                     ^
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:209:19: error: use of undeclared identifier 'MPI_Request'
  static std::set<MPI_Request *> requests;
                  ^
/Users/jenkins/jenkins/workspace/dealii-osx_PR-15572/include/deal.II/base/init_finalize.h:209:32: error: expected expression
  static std::set<MPI_Request *> requests;
                               ^
4 errors generated.
```

**Comment by kronbichler** [comment]
FYI @sloede

**Comment by tamiko** [comment]
We have at least one failing test with:
```
--------------------------------------------------------
An error occurred in line <260> of file </__w/dealii/dealii/source/base/init_finalize.cc> in function
    dealii::InitFinalize::InitFinalize(int&, char**&, const dealii::InitializeLibrary&, unsigned int)
The violated condition was: 
    nth_process_on_host > 0
```

**Review by tamiko — APPROVED** [review]

**Comment by masterleinad** [comment]
This needs to be rebased and probably needs a changelog entry.

**Comment by Rombur** [comment]
I forgot I still had this PR opened. I've rebased it on master and fixed the conflicts.

**Review by kronbichler — APPROVED** [review]
I think it is time to move forward with this PR. I like this path, and would opt to merge it soon. Unless someone else speaks up, I will merge this tomorrow.

File: include/deal.II/base/init_finalize.h [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+{
+  return static_cast<InitializeLibrary>(static_cast<unsigned int>(f1) |
+                                        static_cast<unsigned int>(f2));
+}
+
+
+
+/**
+ * Global operator which sets the bits from the second argument also in the
+ * first one.
```
Comment (masterleinad) [inline include/deal.II/base/init_finalize.h:83]:
```suggestion
 * Global operator which returns an object in which all bits are set that are
 * set both in the first and second argument.
```

**Review by masterleinad — CHANGES_REQUESTED** [review]
Can we add a test that doesn't initialize all libraries?

File: include/deal.II/base/mpi.h [hunk 2]
```diff
@@ -1137,78 +1135,7 @@ namespace Utilities
...
-    private:
-      /**
-       * Requests to MPI_Wait before finalizing
-       */
-      static std::set<MPI_Request *> requests;
-
-#ifdef DEAL_II_WITH_PETSC
-      bool finalize_petscslepc;
-#endif
+      ~MPI_InitFinalize() = default;
```
[the remark this answers is not in the document]
Comment (Rombur) [reply include/deal.II/base/mpi.h:1138]: It was because of a warning from nvcc but I don't remember which version of nvcc I was using.

**Comment by Rombur** [comment]
@masterleinad `ping`

**Review by masterleinad — APPROVED** [review]

**Pull request merged**